crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -334,21 +336,38 @@ where
         rrset_proofs.insert((name, record_type), proof);
     }
 
+    // assign TTL for all records that match name and record type.
+    let mut rrsig_ttls: HashMap<(Name, RecordType), u32> = HashMap::new();
+
     // set the proofs of all the records, all records are returned, it's up to downstream users to check for correctness
     let mut records = records;
     for record in &mut records {
         // the RRSIG used to validate a record inherits the outcome of the validation
         // for RRSIGs, we need to use their TYPE_COVERED field instead of `RecordType::RRSIG` as the
         // `RecordType` key in `rrset_proofs`
-        let record_type = if let RData::DNSSEC(DNSSECRData::RRSIG(rrsig)) = record.data() {
-            rrsig.type_covered()
+        //
+        // The adjusted TTL is determined by RRSIG + RR record it covers.
+        let (record_type, new_ttl) = if let RData::DNSSEC(DNSSECRData::RRSIG(rrsig)) = record.data()
+        {
+            (rrsig.type_covered(), authenticated_ttl(rrsig, record, current_time))
         } else {
-            record.record_type()
+            (record.record_type(), record.ttl())
         };
 
-        rrset_proofs
-            .get(&(record.name().clone(), record_type))
-            .map(|proof| record.set_proof(*proof));
+        if let Some(proof) = rrset_proofs.get(&(record.name().clone(), record_type)) {
+            record.set_proof(*proof);
+            if proof.is_secure() {
+                record.set_ttl(new_ttl);
+                rrsig_ttls.insert((record.name().clone(), record_type), new_ttl);
+            }
+        }
+    }
+
+    // update TTL for all associated RRs that the RRSIG record covers.
+    for record in &mut records {
```

Comment:
I don't think this loop, or the `rrsig_ttls` hash map, is needed because `record.set_ttl` is already called in the preceding loop?

Reply:
or at the very least

> rrsig_ttls.insert((record.name().clone(), record_type), new_ttl);

this seems like it could do a double insertion per key resulting in the old value to be lost. meaning which values are seen in the next iteration is going to be dependent on the order of the items in `records`

maybe you want to only add entries to the `rrsig_ttls` dictionary if `record` was an RRSIG?

Reply:
The issue here is that the `ttl` is determined by the relational pair of `RRSIG` & `RR`. In the previous loop the authenticated ttl is only available when the `RRSIG` is found (also to get its covered type). Setting `record.set_ttl(new_ttl);` in the previous loop is redundant for RRs, but required for RRSIG. Then the `ttl` is stored in the HashMap `rrsig_ttls` by `Name` & `RecordType` to be available to all the covering RRs. If I omit the last loop, the `ttl` is only set in the RRSIG, but not its RR.

Reply:
I know that this function is unnecessarily complex, because for one it hides inner relations between records (in variable `records`), or that the function does not determine **which** RRSIG covers the RRset.

Reply:
I had the same thought as @japaric here. Could we track the TTL with the `rrset_proofs` map? That has only `Proof` now, but it could be `(Proof, TTL)`, I think? And it would consolidate the `ttl` calculation. This would remove the need for this loop and the additional hashmap I think...